contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,99 @@
+use crate::error::ContractError;
+use soroban_sdk::{contract, contractimpl, Address, BytesN, Env, String, Symbol, Val};
+
+#[contract]
+pub struct Upgrader;
+
+#[contractimpl]
+impl Upgrader {
+    pub fn __constructor(_env: Env) {}
+    pub fn upgrade(
+        env: Env,
+        contract_address: Address,
+        new_version: String,
+        new_wasm_hash: BytesN<32>,
+        migration_data: soroban_sdk::Vec<Val>,
+    ) -> Result<(), ContractError> {
+        assert_new_version_is_different(&env, &contract_address, &new_version)?;
+
+        env.invoke_contract::<()>(
+            &contract_address,
+            &upgrade_fn(&env),
+            soroban_sdk::vec![&env, new_wasm_hash.into()],
+        );
+        env.invoke_contract::<()>(&contract_address, &migrate_fn(&env), migration_data);
+
+        assert_new_version_matches_expected(&env, &contract_address, &new_version)
+    }
+}
+
+fn migrate_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "migrate")
+}
+
+fn upgrade_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "upgrade")
+}
+
+fn version_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "version")
+}
+
+fn assert_new_version_is_different(
+    env: &Env,
+    contract_address: &Address,
+    new_version: &String,
+) -> Result<(), ContractError> {
+    let cur_version: String = env.invoke_contract(
+        &contract_address,
+        &version_fn(&env),
+        soroban_sdk::vec![&env],
+    );
+
+    if cur_version != *new_version {
+        Ok(())
+    } else {
+        Err(ContractError::SameVersion)
+    }
+}
+
+fn assert_new_version_matches_expected(
+    env: &Env,
+    contract_address: &Address,
+    new_version: &String,
+) -> Result<(), ContractError> {
+    let cur_version: String = env.invoke_contract(
+        &contract_address,
+        &version_fn(&env),
+        soroban_sdk::vec![&env],
+    );
+
+    if cur_version == *new_version {
+        Ok(())
+    } else {
+        Err(ContractError::VersionMismatch)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::contract::{Upgrader, UpgraderClient};
+    use axelar_gateway::AxelarGateway;
+    use soroban_sdk::{BytesN, Env, String};
+
+    const _WASM: &[u8] = include_bytes!("testdata/axelar_gateway.wasm");
```

Comment:
`testdata/axelar_gateway.optimized.wasm` or `testdata/axelar_gateway.wasm`?

Reply:
tests are still completely in progress, I just wanted to check in what I had on Friday to showcase the logic